docs/architecture/adr-002-lite-client.md

```diff
@@ -0,0 +1,454 @@
+
+```rust
+pub trait Header {
+    fn height(&self) -> Height;
+    fn bft_time(&self) -> Time;
+    fn validators_hash(&self) -> Hash;
+    fn next_validators_hash(&self) -> Hash;
+
+    fn hash(&self) -> Hash;
+}
+```
+
+#### Validator
+
+A validator contributes a positive voting power if a message was correctly signed by it,
+otherwise it contributes 0. We could represent this with a single method that
+returns either 0 or the voting power, but it's probably clearer with two methods:
+
+```rust
+pub trait Validator {
+    fn power(&self) -> u64;
+    fn verify_signature(&self, sign_bytes: &[u8], signature: &[u8]) -> bool;
+}
+```
+
+#### Vote
+
+A vote corresponds to the message signed by a particular validator for a particular commit.
+Since it is expected to be for a particular commit, it does not need to include
+that information explicitly (eg. the height, the round, the block ID, etc.).
+Instead, from the verifier's perspective, all that information can be contained implicitly within the message
+bytes that are being signed. When those message bytes and the signature are passed to 
+`validator.verify_signature` for the corresponding validator, it should return true:
+
+```rust
+pub trait Vote {
+    fn validator_id(&self) -> ValID;
+    fn sign_bytes(&self) -> &[u8];
+    fn signature(&self) -> &[u8];
+}
+```
+
+#### Validator Set
+
+A validator set is a collection of validators that we will want to iterate over
+to check if each validator signed their vote correctly.
+We also need to know its hash, so we can check it against what's in a header. 
+We use an associated type for the validator:
+
+```rust
+pub trait ValidatorSet {
+    type Validator: Validator;
+
+    fn hash(&self) -> Hash;
+    fn into_vec(&self) -> Vec<Self::Validator>;
+}
+```
+
+It may seem more natural to use an `iter` method here than `into_vec`, since
+ultimately we only need to iterate over the underlying associated type.
+However, this turns out to be much more difficult than expected due to the lack
+of size information in the `Iter` trait. It's possible to use another associated
+type, eg. `type ValidatorIter: ExactSizeIterator<Item = Self::Validator>;`, but it just
+complicates things significantly. If this becomes a performance issue (ie. we
+end up having to copy the vector of validators), we can address it then.
+
+#### Commit
+
+A commit is a collection of votes that we will want to iterate over so we can
+verify each vote.
+We also need to check what block this commit is for (ie. the hash of the
+header). 
+Since the commit is for a particular header, we can require all votes to be for that header, 
+and otherwise ignore them. Again, we use an associated type:
+
+```rust
+pub trait Commit {
+    type Vote: Vote;
+
+    fn header_hash(&self) -> Hash;
+    fn into_vec(&self) -> Vec<Option<Self::Vote>>;
+}
+```
+
+Note the `Option` here. When it is `None`, it indicates that either:
+
+- there was no vote from this validator
+- the validator voted nil
```

Comment:
A link to what "voted nil" means would be helpful if someone wanted to use this library, is mainly a Rust dev (who doesn't often deal with `nil`s), and not familiar with this edge case in the tendermint side? Looks like the best description is actually also in adr25?